File: php/4pay_client.php
<?php

// 引入 hashlib 库
require_once 'path/to/hashlib.php';

// 定义常量
define('PAY_HOST', 'https://post.coinpay.io');
define('ACCESS_KEY', '654321');

// 定义 URL
$create_withdraw_order_url = PAY_HOST . '/coin/pay/withdraw/order/create';
$query_withdraw_order_url = PAY_HOST . '/coin/pay/recharge/order/status';

// 定义 WithdrawOrder 类
class WithdrawOrder
{
    private $uid;
    private $merchant_order_no;
    private $currency_coin_name;
    private $channel_code;
    private $amount;
    private $bank_code;
    private $memo;

    public function __construct($uid, $merchant_order_no, $currency_coin_name, $channel_code, $amount, $bank_code, $memo)
    {
        $this->uid = $uid;
        $this->merchant_order_no = $merchant_order_no;
        $this->currency_coin_name = $currency_coin_name;
        $this->channel_code = $channel_code;
        $this->amount = $amount;
        $this->bank_code = $bank_code;
        $this->memo = $memo;
    }

    public function toArray()
    {
        return [
            'uid' => $this->uid,
            'merchantOrderNo' => $this->merchant_order_no,
            'currencyCoinName' => $this->currency_coin_name,
            'channelCode' => $this->channel_code,
            'amount' => $this->amount,
            'bankCode' => $this->bank_code,
            'memo' => $this->memo
        ];
    }
}

// 定义 QueryWithdrawOrder 类
class QueryWithdrawOrder
{
    private $uid;
    private $merchant_order_no;

    public function __construct($uid, $merchant_order_no)
    {
        $this->uid = $uid;
        $this->merchant_order_no = $merchant_order_no;
    }

    public function toArray()
    {
        return [
            'uid' => $this->uid,
            'merchantOrderNo' => $this->merchant_order_no
        ];
    }
}

// 定义 BaseResponse 类
class BaseResponse
{
    public $code;
    public $msg;
    public $data;

    public function __construct($code, $msg, $data)
    {
        $this->code = $code;
        $this->msg = $msg;
        $this->data = $data;
    }

    public function __toString()
    {
        return "BaseResponse(code={$this->code}, msg={$this->msg})";
    }
}

// 定义 WithdrawOrderResponse 类
class WithdrawOrderResponse extends BaseResponse
{
    public $uid;
    public $merchant_order_no;
    public $record_id;
    public $amount;
    public $url;
    public $qrcode;
    public $bank_id;
    public $account_number;
    public $bank_owner;

    public function __construct($code, $msg, $data)
    {
        parent::__construct($code, $msg, $data);
        $this->uid = $data['uid'];
        $this->merchant_order_no = $data['merchantOrderNo'];
        $this->record_id = $data['recordId'];
        $this->amount = $data['amount'];
        $this->url = $data['url'];
        $this->qrcode = $data['qrcode'];
        $this->bank_id = $data['bankId'];
        $this->account_number = $data['accountNumber'];
        $this->bank_owner = $data['bankOwner'];
    }

    public function __toString()
    {
        return "WithdrawOrderResponse(code={$this->code}, msg={$this->msg}, uid={$this->uid}, merchant_order_no={$this->merchant_order_no}, record_id={$this->record_id}, amount={$this->amount}, url={$this->url}, qrcode={$this->qrcode}, bank_id={$this->bank_id}, account_number={$this->account_number}, bank_owner={$this->bank_owner})";
    }
}

// 定义 QueryWithdrawOrderResponse 类
class QueryWithdrawOrderResponse extends BaseResponse
{
    public $merchant_order_no;
    public $uid;
    public $status;
    public $real_amount;
    public $order_amount;
    public $fee;
    public $record_id;

    public function __construct($code, $msg, $data)
    {
        parent::__construct($code, $msg, $data);
        $this->merchant_order_no = $data['merchantOrderNo'];
        $this->uid = $data['uid'];
        $this->status = $data['status'];
        $this->real_amount = $data['realAmount'];
        $this->order_amount = $data['orderAmount'];
        $this->fee = $data['fee'];
        $this->record_id = $data['recordId'];
    }

    public function __toString()
    {
        return "QueryWithdrawOrderResponse(code={$this->code}, msg={$this->msg}, merchant_order_no={$this->merchant_order_no}, uid={$this->uid}, status={$this->status}, real_amount={$this->real_amount}, order_amount={$this->order_amount}, fee={$this->fee}, record_id={$this->record_id})";
    }
}

// 创建订单
function create_withdraw_order($withdraw_order)
{
    $params = $withdraw_order->toArray();
    $params['sign'] = sign($params, ACCESS_KEY);
    $response = postJSON($create_withdraw_order_url, $params);
    $data = json_decode($response, true);
    return new WithdrawOrderResponse($data['code'], $data['msg'], $data['data']);
}

// 查询订单
function query_withdraw_order($query_order)
{
    $params = $query_order->toArray();
    $params['sign'] = sign($params, ACCESS_KEY);
    $response = postJSON($query_withdraw_order_url, $params);
    $data = json_decode($response, true);
    return new QueryWithdrawOrderResponse($data['code'], $data['msg'], $data['data']);
}

// 计算签名
function sign($params, $access_key)
{
    ksort($params);
    $str = '';
    foreach ($params as $k => $v) {
        $str .= $k . '=' . $v . '&';
    }
    $str .= 'key=' . $access_key;
    return md5($str);
}

// 发送 POST 请求
function postJSON($url, $data)
{
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($ch, CURLOPT_POST, 1);
    curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
    curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/json']);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}

File: python/4pay_client.py
import hashlib
from typing import Dict, Final

import requests


def sign(params: Dict[str, str], access_key: str) -> str:
    """
    Calculates a signature for given params and access key.

    :param params: A dictionary of parameters to sign
    :param access_key: The key to use for signing
    :return: A signature as a hexadecimal string
    """
    return hashlib.md5(
        "&".join(f"{k}={v}" for k, v in sorted(params.items()))
        .encode()
        + access_key.encode()
    ).hexdigest()


pay_host = "https://post.coinpay.io"
access_key = "654321"

create_withdraw_order_url: Final[str] = f"{pay_host}/coin/pay/withdraw/order/create"
query_withdraw_order_url: Final[str] = f"{pay_host}/ /coin/pay/recharge/order/status"


class WithdrawOrder:
    def __init__(self, uid: int, merchant_order_no: str, currency_coin_name: str, channel_code: str, amount: float,
                 bank_code: str, memo: str):
        self.uid = uid
        self.merchant_order_no = merchant_order_no
        self.currency_coin_name = currency_coin_name
        self.channel_code = channel_code
        self.amount = amount
        self.bank_code = bank_code
        self.memo = memo

    def to_dict(self):
        return {
            "uid": self.uid,
            "merchantOrderNo": self.merchant_order_no,
            "currencyCoinName": self.currency_coin_name,
            "channelCode": self.channel_code,
            "amount": self.amount,
            "bankCode": self.bank_code,
            "memo": self.memo
        }

class QueryWithdrawOrder:
    def __init__(self, uid: int, merchant_order_no: str):
        self.uid = uid
        self.merchant_order_no = merchant_order_no
    def to_dict(self):
        return {
            "uid": self.uid,
            "merchantOrderNo": self.merchant_order_no,
        }

class BaseResponse:
    def __init__(self, code: int, msg: str, data: Dict[str, str]):
        self.code = code
        self.msg = msg
        self.data = data

    def __str__(self):
        return f"BaseResponse(code={self.code}, msg={self.msg})"


class WithdrawOrderResponse(BaseResponse):
    def __init__(self, code: int, msg: str, data: Dict[str, str]):
        super().__init__(code, msg, data)
        self.uid = data.get("uid")
        self.merchant_order_no = data.get("merchantOrderNo")
        self.record_id = data.get("recordId")
        self.amount = data.get("amount")
        self.url = data.get("url")
        self.qrcode = data.get("qrcode")
        self.bank_id = data.get("bankId")
        self.account_number = data.get("accountNumber")
        self.bank_owner = data.get("bankOwner")

    def __str__(self):
        return f"WithdrawOrderResponse(code={self.code}, msg={self.msg}, uid={self.uid}, merchant_order_no={self.merchant_order_no}, record_id={self.record_id}, amount={self.amount}, url={self.url}, qrcode={self.qrcode}, bank_id={self.bank_id}, account_number={self.account_number}, bank_owner={self.bank_owner})"

class QueryWithdrawOrderResponse(BaseResponse):
    def __init__(self, code: int, msg: str, data: Dict[str, str]):
        super().__init__(code, msg, data)
        self.merchant_order_no = data.get("merchantOrderNo")
        self.uid = data.get("uid")
        self.status = data.get("status")
        self.real_amount = data.get("realAmount")
        self.order_amount = data.get("orderAmount")
        self.fee = data.get("fee")
        self.record_id = data.get("recordId")

    def __str__(self):
        return f"QueryWithdrawOrderResponse(code={self.code}, msg={self.msg}, merchant_order_no={self.merchant_order_no}, uid={self.uid}, status={self.status}, real_amount={self.real_amount}, order_amount={self.order_amount}, fee={self.fee}, record_id={self.record_id})"

# 创建订单
def create_withdraw_order(withdraw_order: WithdrawOrder) -> WithdrawOrderResponse:
    params = withdraw_order.to_dict()
    params["sign"] = sign(params, access_key)
    with requests.Session() as session:
        response = session.post(create_withdraw_order_url, data=params)
    data: Dict[str, any] = response.json()
    return WithdrawOrderResponse(data["code"], data["msg"], data["data"])

# 查询订单
def query_withdraw_order(query_order: QueryWithdrawOrder) -> QueryWithdrawOrderResponse:
    """
    Queries the withdrawal order.

    :param query_order: The order to query, as a QueryWithdrawOrder instance.
    :return: The response for the query, as a QueryWithdrawOrderResponse instance.
    """
    params: Dict[str, str] = query_order.to_dict()
    params["sign"] = sign(params, access_key)
    with requests.Session() as s:
        response: requests.Response = s.post(query_withdraw_order_url, data=params)
    data: Dict[str, any] = response.json()
    return QueryWithdrawOrderResponse(data["code"], data["msg"], data["data"])